tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python
+#############################################################################
+# Copyright (c) 2020 Balabit
+# Copyright (c) 2021 Xiaoyu Qiu
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 2 as published
+# by the Free Software Foundation, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# As an additional exemption you are allowed to compile & link against the
+# OpenSSL libraries as published by the OpenSSL project. See the file
+# COPYING for details.
+#
+#############################################################################
+import pytest
+
+
+test_parameters_raw = [
+    ("foo", '''""''', '''"(?<DN>foo)"''', "", "${DN}", True, "foo"),
```

Comment:
First of all thanks for the Light testcases.
My comment for different testcases: 
- Can you please add a more basic testcase, when 'patterns' is a simple pattern match without any reference saving: \<DN\>. In this case you can use a custom prefix value to store the result.
- Can you please give a better name for DN. If I am right it is just custom macro name
- Could you please give a testcase for this statement also? "use multiple regular expressions, these should be processed in order, the first successfully matching one causes the processing to stop"
- Could you please add a testcase for the following: "Error parsing regexp-parser, compiling the regexp failed", in this case syslog-ng should not started

Reply:
Thank you for your comments, I will work on this soon.

Reply:
I have updated the commits with the following changes:
* A new test case has been added, whose `id` is `match_without_reference_saving`
* I change the reference name to `<key>`
* A new test case has been added, whose `id` is `match_multiple_regular_expressions`
* A new test case has been added, whose `id` is `regular_expression_compile_error`
* A new test case has been added, whose `id` is `check_MSG_macro`
* A new test case has been added, whose `id` is `unmatch_case`. I have made it available to use 'DriverStatsHandler' in Parser class and extended the API in 'DriverStatsHandler' for parsers. The details can be found in commits [a16d34](https://github.com/syslog-ng/syslog-ng/pull/3702/commits/a16d34f3e3fef3840c040bab565834b3fad34925) and [4b2b90](https://github.com/syslog-ng/syslog-ng/pull/3702/commits/4b2b90a2548537a690cd0dba78e06b6bb7634334)

Reply:
Thanks for the changes, I will check it soon.

Reply:
Thanks for the changes again, they are very good. I have only a few minor review comments.